src/system/SystemObject.h

```diff
@@ -193,6 +205,59 @@ class ObjectPool
 #endif
 };
 
+template <class T, unsigned int N>
+inline void ObjectPool<T, N>::Init()
+{
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+    unsigned int oldCount = __sync_fetch_and_add(&this->mRefCount, 1);
+
+    if (oldCount == 0)
```

Comment:
I see how the atomic operations ensure correctness for `mRefCount`.  However, how do we ensure that the pool is actually initialized after the second call to `Init` returns?  I don't think we actually can.

It seems like this is unsafe unless all of the pools are carefully initialized at the start of the program before anything actually uses them.  That seems to somewhat defeat the purpose of using atomics for `mRefCount`, or having the reference counting scheme at all.

Reply:
Given that we're moving to purely external synchronization, I would just leave it out unless there's an actual race (and I don't see how there can be; I think we'll have bigger problems if somebody tries to initialize the sdk twice), but if necessary, prefer std::atomic to gcc instrinsics.